### PR TITLE
Convert course instances sync sproc to regular query

### PR DESCRIPTION
**Pull request by jonatanschroeder**
# Description

As in title. Most of the functionality was based on regular SQL queries, so there were direct translations in a good chunk of the code. A few minor changes:
* The process was split into two: one part for inserting/deleting records and updating the UUID/short name, and the other to update the rest. This is because Postgres doesn't allow one CTE to update a record that was inserted in another CTE in the same query.
* ~~The original sproc included some assertions to ensure the data was correct. I did not include those assertions.~~
* For access rules uids, there was a check for `= NULL::JSON` which always failed, so I removed it altogether. The sproc that handles the conversion to string arrays already takes care of null anyway.

# Testing

CI already does a good chunk of the testing, so local testing mostly complemented those tests. Course instances are created, updated and deleted as expected.

### CODE REVIEW

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts [hunk 1]
```diff
@@ -116,26 +116,45 @@ export async function sync(
...
         shortName,
-        courseInstance.uuid,
+        courseInstanceData.courseInstance.uuid,
         // This enrollment code is only used for inserts, and not used on updates
         await uniqueEnrollmentCode(),
+      ]),
+    ),
+  );
+
+  const shortNameToIdMapping = await sqldb.queryScalar(
```
Comment (mwest1066) [inline apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: Should this be wrapped in a `runInTransactionAsync()` with the `sync_course_instances_update` query below?

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts [hunk 2]
```diff
@@ -116,26 +116,45 @@ export async function sync(
...
+    sql.sync_course_instances_insert_delete,
+    { course_instances_data: courseInstanceIdentityParams, course_id: courseId },
+    z.record(z.string(), IdSchema),
+  );
+
+  const courseInstanceGeneralParams = Object.entries(courseData.courseInstances).map(
+    ([shortName, courseInstanceData]) => {
+      const courseInstanceId = shortNameToIdMapping[shortName];
+      if (!courseInstanceId) {
+        throw new Error(
```
Comment (mwest1066) [inline apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: Along with this error, should we retain the consistency checks from the old sproc?

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts (hunk 1 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: Done in 4c5fd2f0853a133ebc38417967b1afe0d1eb4c8d. I also added the uniqueEnrollmentCode calls into the transaction with #14359 in mind, though this does not fix the issue yet.

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts (hunk 2 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: The other consistency checks in the old sproc would require an additional query to retrieve the list of course instances for the course unjoined by the sync data. I'm not convinced they are necessary, but if you have a reason to believe they are needed I can be persuaded. There are basically three checks:

- All course instances retrieved from disk have been added/updated in sync (covered by the check above).
- All non-deleted course instances in the DB have an equivalent in disk.
- All non-deleted course instances in the DB match the UUID of the corresponding course instance in disk.

Comment (reteps) [reply apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: My understanding is that checks 2 and 3 are tested pretty thoroughly in the sync code, at least according to Claude. I also tried to look for gaps in test coverage, but only found this case: "Syncing with zero course instances". Additionally, the sync tests don't test enrollment code generation, only `enroll.test.ts` does.

File: apps/prairielearn/src/sync/fromDisk/courseInstances.sql [hunk 3]
```diff
@@ -13,3 +13,243 @@ FROM
...
+      -- which is currently true. Mainly, this ensures that the tests are
+      -- deterministic. We specifically use C collation to ensure that "Sp15"
+      -- ends up before "public".
+      -- See: https://github.com/PrairieLearn/PrairieLearn/pull/12037 and follow-up PRs.
+      src_short_name COLLATE "C" ASC
+    RETURNING
+      dest.short_name AS src_short_name,
+      dest.id AS inserted_dest_id
+  )
+  -- At this point, there will be exactly one non-deleted row for
```
Comment (reteps) [inline apps/prairielearn/src/sync/fromDisk/courseInstances.sql]: some of these comments make less sense here. "At this point" -- this isn't a sproc anymore, this comment should be lifted to typescript presumably to continue the narration?

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts (hunk 2 above)
Comment (reteps) [reply apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: Er, I am siding with @jonatanschroeder here -- tests are validating these cases, so the internal sanity checks don't seem required.

Comment (mwest1066) [reply apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: While tests are of course also good, I'm generally in favor of having internal consistency checks for critical functionality, especially things like sync that maintain data integrity. I've certainly had similar checks catch bugs in the past, even when tests were fully passing. This is because tests only ever cover a subset of all possible conditions and it can be easy to miss the weird combination of inputs that causes a failure. For example, we typically don't have test cases that cover race conditions, second calls to the same function after an earlier failure, etc. Pre/post consistency checks are one way of trying to cover such cases.

While I agree that there will be a slight speed penalty from retaining the checks in this case, I'm still in favor of keeping them.

Comment (jonatanschroeder) [reply apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: Checks added in c30800f92f14d3ddbf3de8de63d836285d725859.

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts [hunk 4]
```diff
@@ -160,3 +165,62 @@ export async function sync(
...
+    CourseInstanceSchema.pick({ short_name: true, uuid: true }),
+  );
+  const shortNamesInDb = courseInstancesForConsistencyCheck.map((ci) => ci.short_name);
+  const shortNamesInDisk = Object.keys(courseInstances);
+
+  const duplicateShortNames = Object.entries(countBy(shortNamesInDisk, (name) => name))
+    .filter(([, count]) => count > 1)
+    .map(([shortName]) => shortName);
+  if (duplicateShortNames.length > 0) {
+    throw new AugmentedError(
```
Comment (reteps) [inline apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: I'm not sure if this gets surfaced as an error in Sentry, which we should make sure it does. I'll ask about this in standup.

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts [hunk 5]
```diff
@@ -116,26 +120,114 @@ export async function sync(
...
+
+    const duplicateShortNames = Object.entries(countBy(shortNamesInDisk, (name) => name))
+      .filter(([, count]) => count > 1)
+      .map(([shortName]) => shortName);
+    if (duplicateShortNames.length > 0) {
+      throw new AugmentedError(
+        'Assertion: Duplicate course instance short names found in disk data',
+        { data: { courseId, duplicateShortNames } },
+      );
+    }
```
Comment (nwalters512) [inline apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: I'm not sure what this is even doing. `shortNamesInDisk` is derived directly from `Object.keys(...)`, and an object by definition cannot have duplicate keys. @jonatanschroeder what was the goal here?

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts (hunk 4 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: I don't think they are, from my test (creating a bug to reproduce the validation error) the error is captured by the server job handler and printed, causing the job to fail without (to my knowledge) sending it to Sentry. I added sentry reporting in df49e37183b1c1a1af4b13eeee4179d3bb4fa6e3.

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts (hunk 5 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: Ah, my bad, this was supposed to come from DB, not disk. Fix coming.

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts (hunk 4 above)
Comment (nwalters512) [reply apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: They do not; we expect server jobs to fail from time to time for legit reasons and don't automatically send any errors to Sentry.

In this case, we probably should. I think caller of `courseInstanceConsistencyCheck` could do a `Sentry.captureException(...)` before allowing the error to continue back up the stack.

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts (hunk 5 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: Fixed in e1b3093b65c932b21fc7b777c570c66015dbc7f3. I added this validation because short name is not a unique value in course instances, so this ensures that short names are updated instead of re-inserted.

File: apps/prairielearn/src/sync/fromDisk/courseInstances.ts (hunk 4 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/sync/fromDisk/courseInstances.ts]: Noted. As mentioned, this is done.
